e2e/vmcompose/provider.go

```diff
@@ -123,26 +123,40 @@ func (p *Provider) Setup() error {
 	return nil
 }
 
+// Upgrade copies some of the local e2e generated artifacts to the VMs and starts the docker-compose services.
 func (p *Provider) Upgrade(ctx context.Context) error {
 	log.Info(ctx, "Upgrading docker-compose on VMs", "image", p.Testnet.UpgradeVersion)
 
 	var filePaths []string
-	// add all artifacts for each node
+	addFile := func(dir string, file string) {
+		filePaths = append(filePaths, filepath.Join(p.Testnet.Dir, dir, file))
+	}
+
+	// TODO(corver): Also upgrade long-lived keys if changed
+	// - validator: ensure privval_state.json and voter_state.json doesn't complain
+
+	// Include halo config
 	for _, node := range p.Testnet.Nodes {
-		nodeDir := filepath.Join(p.Testnet.Dir, node.Name)
+		addFile(node.Name, "config/halo.toml")
+		addFile(node.Name, "config/config.toml")
+		addFile(node.Name, "config/network.json")
+		addFile(node.Name, "config/jwtsecret")
```

Comment:
Can the secret change?

Reply:
it is randomly generated each time, so yes. Might as well rotate it.